### PR TITLE
Dircon Rework

### PR DESCRIPTION
**Pull request by mposa**
This is not anywhere near ready to merge yet, but it'd be good to get feedback on the `DirconMode` API, as Dircon programs will be entirely specified by one or more `DirconMode` objects. Comments should focus on dircon_mode.h and, to a lesser extent, dircon_mode.cc

### CODE REVIEW

File: systems/trajectory_optimization/dircon/dircon_mode.h [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+/// DirconMode is the primary class for defining constrained collocation
+/// (DIRCON) trajectory optimization problems. Each DirconMode object refers
+/// to a single hybrid mode of the optimization.  
+template <typename T>
+class DirconMode {
+ public:
+
+  /// @param evaluators The set of constraints, active and full.
+  /// @param num_knotpoints The total number of knot points.
```
Comment (yminchen) [inline systems/trajectory_optimization/dircon/dircon_mode.h:26]: I think `num_knot` is probably clear enough

File: systems/trajectory_optimization/dircon/dircon_mode.h [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  /// KinematicEvaluator.
+  const std::unordered_map<int, double>& GetKinAccelerationScale() const {
+    return kin_acceleration_scale_;
+  };
+
+  /// Get the cost on force regularization. Defaults to 1e-4
+  double get_force_regularization() const { return force_regularization_;};
+
+ private:
+  void SetKinScale(std::unordered_map<int, double>* scale_map,
```
Comment (yminchen) [inline systems/trajectory_optimization/dircon/dircon_mode.h:157]: What is the role of this method?

File: systems/trajectory_optimization/dircon/dircon_mode.h [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+      int evaluator_index, int constraint_index, double scale);
+ 
+  const multibody::KinematicEvaluatorSet<T>& evaluators_;
+  int num_knotpoints_;
+  double max_T_;
+  double min_T_;
+  std::set<int> relative_constraints_;
+  std::unordered_map<int, DirconKinConstraintType> reduced_constraints_;
```
Comment (yminchen) [inline systems/trajectory_optimization/dircon/dircon_mode.h:165]: Is this variable used to keep track of the constraint type of the kinematics constraints? If so, the name could be a bit confusing

**Comment by mposa** [comment]
Closing this PR, will reopen when ready for review.

**Pull request closed without merging**